megatron/core/typed_torch.py

```diff
@@ -0,0 +1,31 @@
+"""Utilities for improved type hinting with torch interfaces."""
+
+from collections.abc import Callable
+from typing import Generic, ParamSpec, Protocol, TypeVar
+
+import torch
+
+P = ParamSpec('P')
+R_co = TypeVar('R_co', covariant=True)
+
+
+class _Module(Generic[P, R_co], Protocol):
+    """Protocol allowing us to unwrap `forward`."""
+
+    def forward(self, *args: P.args, **kwargs: P.kwargs) -> R_co: ...
+
+
+def apply_module(m: _Module[P, R_co], *, check_subclass: bool = True) -> Callable[P, R_co]:
```

Comment:
You can use a typing overload here to have check_subclass=True be a static assertion here. Now if check_subclass is True it will require it to follow the _TrueModule Protocol and complain if it doesn't. Mypy, pyright, etc prefer the more specific overload so this should work.

```suggestion
class _TrueModule(torch.nn.Module, _Module, Protocol):
    pass

@typing.overload
def apply_module(m: _TrueModule[P, R_co], *, check_subclass: Literal[True] = True) -> Callable[P, R_co]:
    pass
    
@typing.overload
def apply_module(m: _Module[P, R_co], *, check_subclass: Literal[True] = True) -> typing.Never:
    pass
```

Reply:
Oh wait no I tried this when I originally wrote it - `Protocols` cannot extend non-`Protocol` classes. Unfortunately python typing is fairly limited and doesn't allow for higher-kinded or constrained types :(